exporters/metric/dynamicconfig/push/push.go

```diff
@@ -0,0 +1,197 @@
+	"go.opentelemetry.io/otel/sdk/metric/integrator/simple"
+)
+
+// DefaultPushPeriod is the default time interval between pushes.
+const DefaultPushPeriod = 10 * time.Second
+
+// Controller organizes a periodic push of metric data.
+type Controller struct {
+	lock        sync.Mutex
+	accumulator *sdk.Accumulator
+	provider    *registry.Provider
+	integrator  *simple.Integrator
+	exporter    export.Exporter
+	wg          sync.WaitGroup
+	ch          chan bool
+	period      time.Duration
+	timeout     time.Duration
+	clock       controllerTime.Clock
+	ticker      controllerTime.Ticker
+	notifier    *dynamicconfig.Notifier
+}
+
+// New constructs a Controller, an implementation of metric.Provider,
+// using the provided exporter and options to configure an SDK with
+// periodic collection.
+func New(selector export.AggregationSelector, exporter export.Exporter, opts ...Option) *Controller {
+	c := &Config{
+		Period: DefaultPushPeriod,
+	}
+	for _, opt := range opts {
+		opt.Apply(c)
+	}
+	if c.Timeout == 0 {
+		c.Timeout = c.Period
+	}
+
+	integrator := simple.New(selector, c.Stateful)
+	impl := sdk.NewAccumulator(
+		integrator,
+		sdk.WithResource(c.Resource),
+	)
+	return &Controller{
+		provider:    registry.NewProvider(impl),
+		accumulator: impl,
+		integrator:  integrator,
+		exporter:    exporter,
+		ch:          make(chan bool),
+		period:      c.Period,
+		timeout:     c.Timeout,
+		clock:       controllerTime.RealClock{},
+		notifier:    c.Notifier,
+	}
+}
+
+// SetClock supports setting a mock clock for testing.  This must be
+// called before Start().
+func (c *Controller) SetClock(clock controllerTime.Clock) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.clock = clock
+}
+
+// Provider returns a metric.Provider instance for this controller.
+func (c *Controller) Provider() metric.Provider {
+	return c.provider
+}
+
+// Start begins a ticker that periodically collects and exports
+// metrics with the configured interval.
+func (c *Controller) Start() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.ticker != nil {
+		return
+	}
+
+	if c.notifier != nil {
+		c.notifier.Register(c)
+	}
+
+	c.ticker = c.clock.Ticker(c.period)
+	c.wg.Add(1)
+	go c.run(c.ch)
+}
+
+// Stop waits for the background goroutine to return and then collects
+// and exports metrics one last time before returning.
+func (c *Controller) Stop() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.ch == nil {
+		return
+	}
+
+	close(c.ch)
+	c.ch = nil
+	c.wg.Wait()
+	c.ticker.Stop()
+	c.ticker = nil
+
+	c.tick()
+
+	if c.notifier != nil {
+		c.notifier.Unregister(c)
+	}
+}
+
+// We assume that the only metric schedule is one with an inclusion pattern
+// that includes all metrics
+// TODO: Change later to support metric schedules
+func (c *Controller) OnInitialConfig(config *dynamicconfig.Config) {
+	if len(config.MetricConfig.Schedules) > 0 {
+		c.period = time.Duration(config.MetricConfig.Schedules[0].Period) * time.Second
```

Comment:
Let's do some sanity check for "Period" for example, it should not be 0. We should not apply the new config if it turns out to be illegal.

Also, you don't need to convert it to time.Duration to multiply with time.Second. That is

config.MetricConfig.Schedules[0].Period * time.Second

should just work.

Reply:
I suggest put it into a separate function (like "ValidateSchedule") and reuse it in OnUpdatedConfig.